controllers/runnerpool_controller.go

```diff
@@ -129,128 +130,71 @@ func (r *RunnerPoolReconciler) finalize(ctx context.Context, log logr.Logger, rp
+		d.Labels = mergeMap(d.GetLabels(), labelSet(rp, constants.AppComponentRunner))
+		d.Spec.Selector = &metav1.LabelSelector{
+			MatchLabels: labelSet(rp, constants.AppComponentRunner),
 		}
+		d.Spec.Template.Labels = labelSetForRunnerPod(rp, r.organizationName)
 
-		// Add metrics port
-		container.Ports = append(container.Ports, corev1.ContainerPort{
-			Protocol:      corev1.ProtocolTCP,
-			Name:          constants.RunnerMetricsPortName,
-			ContainerPort: constants.RunnerMetricsPort,
-		})
+		d.Spec.Replicas = pointer.Int32Ptr(rp.Spec.Replicas)
+		d.Spec.Template.Spec.ServiceAccountName = rp.GetRunnerServiceAccountName()
+		d.Spec.Template.Spec.ImagePullSecrets = rp.Spec.Template.ImagePullSecrets
+		d.Spec.Template.Spec.Volumes = rp.Spec.Template.Volumes
 
-		envMap := make(map[string]struct{})
-		for _, v := range container.Env {
-			envMap[v.Name] = struct{}{}
-		}
+		r.addRunnerContainerIfNotExists(d)
+		runnerContainer := r.findRunnerContainer(d)
 
-		if _, ok := envMap[constants.PodNameEnvName]; !ok {
-			container.Env = append(
-				container.Env,
-				corev1.EnvVar{
-					Name: constants.PodNameEnvName,
-					ValueFrom: &corev1.EnvVarSource{
-						FieldRef: &corev1.ObjectFieldSelector{
-							FieldPath: "metadata.name",
-						},
-					},
-				},
-			)
-		}
-		if _, ok := envMap[constants.PodNamespaceEnvName]; !ok {
-			container.Env = append(
-				container.Env,
-				corev1.EnvVar{
-					Name: constants.PodNamespaceEnvName,
-					ValueFrom: &corev1.EnvVarSource{
-						FieldRef: &corev1.ObjectFieldSelector{
-							FieldPath: "metadata.namespace",
-						},
-					},
-				},
-			)
-		}
-		if _, ok := envMap[constants.RunnerOrgEnvName]; !ok {
-			container.Env = append(
-				container.Env,
-				corev1.EnvVar{
-					Name:  constants.RunnerOrgEnvName,
-					Value: r.organizationName,
-				},
-			)
-		}
-		if _, ok := envMap[constants.RunnerRepoEnvName]; !ok {
-			container.Env = append(
-				container.Env,
-				corev1.EnvVar{
-					Name:  constants.RunnerRepoEnvName,
-					Value: rp.Spec.RepositoryName,
-				},
-			)
+		// Update the runner container.
+		if rp.Spec.Template.Image != "" {
+			runnerContainer.Image = rp.Spec.Template.Image
+		} else {
+			runnerContainer.Image = r.runnerImage
 		}
-		if _, ok := envMap[constants.RunnerPoolNameEnvName]; !ok {
-			container.Env = append(
-				container.Env,
-				corev1.EnvVar{
-					Name:  constants.RunnerPoolNameEnvName,
-					Value: rp2.ObjectMeta.Name,
-				},
-			)
-		}
-		if _, ok := envMap[constants.SlackAgentEnvName]; !ok && rp2.Spec.SlackAgentServiceName != nil {
-			container.Env = append(
-				container.Env,
-				corev1.EnvVar{
-					Name:  constants.SlackAgentEnvName,
-					Value: *rp.Spec.SlackAgentServiceName,
-				},
-			)
+		if rp.Spec.Template.ImagePullPolicy != "" {
+			runnerContainer.ImagePullPolicy = rp.Spec.Template.ImagePullPolicy
 		}
+		runnerContainer.SecurityContext = rp.Spec.Template.SecurityContext
```

Comment:
This may also cause infinite updates.
Because pod-security-admission will mutates SecurityContext.

Reply:
Does the pod-security-admission mutate the deployment?

Reply:
I'm sorry.
I mistook it for Pod Resource.